File: src/Tests/Performance/PerfStringMatch/PerfStringMatch.cs
namespace go;

using fmt = fmt_package;
using strings = strings_package;
using time = time_package;

partial class main_package {

internal static slice<@string> lines = new @string[]{
    "//go:build linux && amd64",
    "// regular comment line",
    "package main",
    "import \"fmt\"",
    "func main() {",
    "\tx := true",
    "}",
    "//go:build windows",
    "var y = false",
    "// another comment"
}.slice();

internal static slice<@string> words = new @string[]{"true", "false", "linux", "windows", "amd64", "arm64", "ignore", "main"}.slice();

internal static nint classify(@string w) {
    var exprᴛ1 = w;
    if (exprᴛ1 == "true"u8) {
        return 1;
    }
    if (exprᴛ1 == "false"u8) {
        return 2;
    }
    if (exprᴛ1 == "linux"u8 || exprᴛ1 == "windows"u8 || exprᴛ1 == "darwin"u8) {
        return 3;
    }
    if (exprᴛ1 == "amd64"u8 || exprᴛ1 == "arm64"u8) {
        return 4;
    }

    return 0;
}

internal static @string kindName(nint k) {
    switch (k) {
    case 1 or 2: {
        return "bool"u8;
    }
    case 3: {
        return "goos"u8;
    }
    case 4: {
        return "goarch"u8;
    }}

    return "word"u8;
}

internal static nint run(nint n) {
    nint total = 0;
    var counts = new map<@string, nint>{};
    for (nint i = 0; i < n; i++) {
        @string line = lines[i % len(lines)];
        if (strings.HasPrefix(line, "//go:build"u8)){
            counts["build"u8]++;
            total += len(line);
        } else 
        if (strings.HasPrefix(line, "// "u8)) {
            counts["comment"u8]++;
        }
        @string w = words[i % len(words)];
        nint k = classify(w);
        total += k;
        total += len(kindName(k));
    }
    total += counts["build"u8] * 3 + counts["comment"u8];
    return total;
}

internal static void Main() {
    var start = time.Now().UnixNano();
    nint total = run(20000000);
    var elapsed = time.Now().UnixNano() - start;
    fmt.Println((@string)"checksum:"u8, total);
    fmt.Println((@string)"elapsed_ns:"u8, elapsed);
}

} // end main_package

File: src/tests/Performance/PerfStringMatch/PerfStringMatch.go
package main

import (
	"fmt"
	"strings"
	"time"
)

// Literal-string hot paths: switch-on-string dispatch, literal-prefix scanning,
// literal returns, and literal map keys -- the shapes the tiered @string literal
// optimizations target in converted C# (span comparison operators, u8 casts,
// hoisted literals). Go allocates nothing at any of these sites.

var lines = []string{
	"//go:build linux && amd64",
	"// regular comment line",
	"package main",
	"import \"fmt\"",
	"func main() {",
	"\tx := true",
	"}",
	"//go:build windows",
	"var y = false",
	"// another comment",
}

var words = []string{"true", "false", "linux", "windows", "amd64", "arm64", "ignore", "main"}

func classify(w string) int {
	switch w {
	case "true":
		return 1
	case "false":
		return 2
	case "linux", "windows", "darwin":
		return 3
	case "amd64", "arm64":
		return 4
	}
	return 0
}

func kindName(k int) string {
	switch k {
	case 1, 2:
		return "bool"
	case 3:
		return "goos"
	case 4:
		return "goarch"
	}
	return "word"
}

func run(n int) int {
	total := 0
	counts := map[string]int{}

	for i := 0; i < n; i++ {
		line := lines[i%len(lines)]
		if strings.HasPrefix(line, "//go:build") {
			counts["build"]++
			total += len(line)
		} else if strings.HasPrefix(line, "// ") {
			counts["comment"]++
		}

		w := words[i%len(words)]
		k := classify(w)
		total += k
		total += len(kindName(k))
	}

	total += counts["build"]*3 + counts["comment"]
	return total
}

func main() {
	start := time.Now().UnixNano()

	total := run(20000000)

	elapsed := time.Now().UnixNano() - start
	fmt.Println("checksum:", total)
	fmt.Println("elapsed_ns:", elapsed)
}
